federatedscope/core/trainers/context.py

```diff
@@ -84,24 +25,79 @@ def __getattr__(self, item):
         except KeyError:
             raise AttributeError("Attribute {} is not found".format(item))
 
+    def __init__(self, init_dict=None):
+        if init_dict is not None:
+            super(LifecycleDict, self).__init__(init_dict)
+        self.lifecycles = collections.defaultdict(set)
+
+    def __setattr__(self, key, value):
+        if isinstance(value, CtxVar):
+            self.lifecycles[value.lifecycle].add(key)
+            super(LifecycleDict, self).__setitem__(key, value.obj)
+        else:
+            super(LifecycleDict, self).__setitem__(key, value)
+
+    def clear(self, lifecycle):
+        keys = list(self.lifecycles[lifecycle])
+        for key in keys:
+            if key in self:
+                del self[key]
+            self.lifecycles[lifecycle].remove(key)
+
+
+class Context(LifecycleDict):
+    """Record and pass variables among different hook functions
+    Arguments:
+        model: training model
+        cfg: config
+        data (dict): a dict contains train/val/test dataset or dataloader
+        device: running device
+        init_dict (dict): a dict used to initialize the instance of Context
+        init_attr (bool): if set up the static variables
+    Note:
+        - The variables within an instance of class `Context` can be set/get as an attribute.
+        ```
+        ctx.${NAME_VARIABLE} = ${VALUE_VARIABLE}
+        ```
+        where `${NAME_VARIABLE}` and `${VALUE_VARIABLE}` is the name and value of the variable.
+
+        - To achieve lifecycle management, you can wrap the variable with `CtxVar` and a lifecycle parameter
+        as follows
+        ```
+        ctx.${NAME_VARIABLE} = CtxVar(${VALUE_VARIABLE}, LFECYCLE.BATCH)
```

Comment:
Perhaps we need to add sth. like your mentioned sentence to make users aware that they can also manage the lifecycle by themself if they want, `The class CtxVar makes no difference to the variables, and you don't need to worry about it. Also, you can manage the lifecycle all by yourself.`

Reply:
> Perhaps we need to add sth. like your mentioned sentence to make users aware that they can also manage the lifecycle by themself if they want, `The class CtxVar makes no difference to the variables, and you don't need to worry about it. Also, you can manage the lifecycle all by yourself.`

Modified accordingly.